### PR TITLE
marker-range: handle overlap

### PR DESCRIPTION
**Pull request by Fumesover**
## Description

The usage of the `marker-range` was generating a too broad range.

For example, the following query:

```clojure
@(store/long-range-reduce *db* incrementor 0 :Object ["small-bucket" ""] {::store/marker "files/00000010.txt"})
```

created the following TupleRange, meaning that all the buckets containing "small-bucket*" are listed:

```clojure
#object[com.apple.foundationdb.record.TupleRange 0x6b424252 "[[4, small-bucket, files/00000010.txt],[4, small-buckeu])"]
```

Thanks to the work of @Fumesover, we are now specifying the EndpointType when creating a range.
- For the `marker`, it will use the `EndpointType/CONTINUATION`
- For the `prefix`, it will use the `EndpointType/PREFIX_STRING`

This PR contains a breaking change as it renames the `marker` into `continuation` which matches more closely
the FoundationDB Record Layer terminology.

## Testing done

```
lein test                                                                                                                                                                                    1 ✘  2 task  at 09:45:33   

lein test exoscale.vinyl.cursor-test

lein test exoscale.vinyl.payload-test
{:result true, :num-tests 100, :seed 1654674340748, :time-elapsed-ms 34, :test-var "user-serialization-test"}
{:result true, :num-tests 100, :seed 1654674340788, :time-elapsed-ms 118, :test-var "invoice-serialization-test"}
{:result true, :num-tests 100, :seed 1654674340907, :time-elapsed-ms 14, :test-var "account-serialization-test"}

lein test exoscale.vinyl.scan-test
[main] INFO exoscale.vinyl.store - started store :demostore / testing5066
[main] INFO exoscale.vinyl.demostore - installing test data: 23 records
[ForkJoinPool.commonPool-worker-3] WARN com.apple.foundationdb.record.provider.foundationdb.FDBDatabaseRunnerImpl - Retrying FDB Exception code="1020" curr_attempt="0" delay="7" max_attempts="10" message="Transaction not committed due to conflict with another transaction"
java.util.concurrent.CompletionException: com.apple.foundationdb.FDBException: Transaction not committed due to conflict with another transaction
	at java.base/java.util.concurrent.CompletableFuture.encodeRelay(CompletableFuture.java:368)
	at java.base/java.util.concurrent.CompletableFuture.completeRelay(CompletableFuture.java:377)
	at java.base/java.util.concurrent.CompletableFuture$UniRelay.tryFire(CompletableFuture.java:1097)
	at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:510)
	at java.base/java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:2162)
	at com.apple.foundationdb.NativeFuture.marshalWhenDone(NativeFuture.java:76)
	at java.base/java.util.concurrent.ForkJoinTask$RunnableExecuteAction.exec(ForkJoinTask.java:1395)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:373)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1182)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1655)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1622)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:165)
Caused by: com.apple.foundationdb.FDBException: Transaction not committed due to conflict with another transaction
	at com.apple.foundationdb.NativeFuture.Future_getError(Native Method)
	at com.apple.foundationdb.FutureVoid.getIfDone_internal(FutureVoid.java:35)
	at com.apple.foundationdb.FutureVoid.getIfDone_internal(FutureVoid.java:25)
	at com.apple.foundationdb.NativeFuture.marshalWhenDone(NativeFuture.java:63)
	... 6 more
[ForkJoinPool.commonPool-worker-4] INFO com.apple.foundationdb.record.provider.foundationdb.FDBRecordStore - new record store format_version="7" key_space_path="/environment:19->'testing5066'/schema:'demostore'" meta_data_version="7"
[main] INFO exoscale.vinyl.demostore - ready to serve fdb queries
[main] INFO exoscale.vinyl.demostore - cleaning fdb store

lein test exoscale.vinyl.sql-test
[main] INFO exoscale.vinyl.store - started store :demostore / testing5067
[main] INFO exoscale.vinyl.demostore - installing test data: 23 records
[ForkJoinPool.commonPool-worker-7] WARN com.apple.foundationdb.record.provider.foundationdb.FDBDatabaseRunnerImpl - Retrying FDB Exception code="1020" curr_attempt="0" delay="5" max_attempts="10" message="Transaction not committed due to conflict with another transaction"
java.util.concurrent.CompletionException: com.apple.foundationdb.FDBException: Transaction not committed due to conflict with another transaction
	at java.base/java.util.concurrent.CompletableFuture.encodeRelay(CompletableFuture.java:368)
	at java.base/java.util.concurrent.CompletableFuture.completeRelay(CompletableFuture.java:377)
	at java.base/java.util.concurrent.CompletableFuture$UniRelay.tryFire(CompletableFuture.java:1097)
	at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:510)
	at java.base/java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:2162)
	at com.apple.foundationdb.NativeFuture.marshalWhenDone(NativeFuture.java:76)
	at java.base/java.util.concurrent.ForkJoinTask$RunnableExecuteAction.exec(ForkJoinTask.java:1395)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:373)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1182)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1655)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1622)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:165)
Caused by: com.apple.foundationdb.FDBException: Transaction not committed due to conflict with another transaction
	at com.apple.foundationdb.NativeFuture.Future_getError(Native Method)
	at com.apple.foundationdb.FutureVoid.getIfDone_internal(FutureVoid.java:35)
	at com.apple.foundationdb.FutureVoid.getIfDone_internal(FutureVoid.java:25)
	at com.apple.foundationdb.NativeFuture.marshalWhenDone(NativeFuture.java:63)
	... 6 more
[ForkJoinPool.commonPool-worker-7] INFO com.apple.foundationdb.record.provider.foundationdb.FDBRecordStore - new record store format_version="7" key_space_path="/environment:17->'testing5067'/schema:'demostore'" meta_data_version="7"
[main] INFO exoscale.vinyl.demostore - ready to serve fdb queries
[main] INFO exoscale.vinyl.demostore - cleaning fdb store

lein test exoscale.vinyl.store-test
[main] INFO exoscale.vinyl.store - started store :demostore / testing5068
[main] INFO exoscale.vinyl.demostore - installing test data: 23 records
[ForkJoinPool.commonPool-worker-5] WARN com.apple.foundationdb.record.provider.foundationdb.FDBDatabaseRunnerImpl - Retrying FDB Exception code="1020" curr_attempt="0" delay="9" max_attempts="10" message="Transaction not committed due to conflict with another transaction"
java.util.concurrent.CompletionException: com.apple.foundationdb.FDBException: Transaction not committed due to conflict with another transaction
	at java.base/java.util.concurrent.CompletableFuture.encodeRelay(CompletableFuture.java:368)
	at java.base/java.util.concurrent.CompletableFuture.completeRelay(CompletableFuture.java:377)
	at java.base/java.util.concurrent.CompletableFuture$UniRelay.tryFire(CompletableFuture.java:1097)
	at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:510)
	at java.base/java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:2162)
	at com.apple.foundationdb.NativeFuture.marshalWhenDone(NativeFuture.java:76)
	at java.base/java.util.concurrent.ForkJoinTask$RunnableExecuteAction.exec(ForkJoinTask.java:1395)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:373)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1182)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1655)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1622)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:165)
Caused by: com.apple.foundationdb.FDBException: Transaction not committed due to conflict with another transaction
	at com.apple.foundationdb.NativeFuture.Future_getError(Native Method)
	at com.apple.foundationdb.FutureVoid.getIfDone_internal(FutureVoid.java:35)
	at com.apple.foundationdb.FutureVoid.getIfDone_internal(FutureVoid.java:25)
	at com.apple.foundationdb.NativeFuture.marshalWhenDone(NativeFuture.java:63)
	... 6 more
[ForkJoinPool.commonPool-worker-5] INFO com.apple.foundationdb.record.provider.foundationdb.FDBRecordStore - new record store format_version="7" key_space_path="/environment:36->'testing5068'/schema:'demostore'" meta_data_version="7"
[main] INFO exoscale.vinyl.demostore - ready to serve fdb queries
[main] INFO exoscale.vinyl.demostore - cleaning fdb store

Ran 12 tests containing 82 assertions.
0 failures, 0 errors.
```

### CODE REVIEW

**Review by arnaudgeiser — APPROVED** [review]

**Review by bcachet — APPROVED** [review]
Really impressed :open_mouth: by the quality of your PR
Reproducing case in the tests is top-notch and fix look solid

File: src/exoscale/vinyl/store.clj [hunk 1]
```diff
@@ -343,23 +344,18 @@
...
-       (if (seq prefix)
-         (key-for* txn-context record-type (conj (vec (butlast items)) (inc-prefix prefix)))
-         (key-for* txn-context record-type (conj (vec (drop-last 2 items)) (inc-prefix (nth items (- (count items) 2))))))))
+(defn ^TupleRange continuation-range
+  "Given a prefix and an optional continuation, return a TupleRange catching all
+   elements with the given prefix, starting from continuation if present."
+  [txn-context record-type items continuation]
+  (if continuation
+    (let [tuple-start (key-for* txn-context record-type (conj (vec (butlast items)) continuation))
+          tuple-end   (key-for* txn-context record-type (conj (vec items)))]
```
Comment (pyr) [inline src/exoscale/vinyl/store.clj]: `conj` with a single argument is a very specific `identity` :-)

```suggestion
          tuple-end   (key-for* txn-context record-type (vec items))]
```

File: src/exoscale/vinyl/store.clj [hunk 2]
```diff
@@ -509,8 +505,8 @@
    taken with the accumulator."
   ([db f val record-type items]
    (long-range-reduce db f val record-type items {}))
-  ([db f val record-type items {::keys [marker] :as opts}]
-   (let [range (marker-range db record-type items marker)
+  ([db f val record-type items {::keys [continuation] :as opts}]
```
Comment (pyr) [inline src/exoscale/vinyl/store.clj:508]: `continuation` is a nicer name, but this constitutes a breaking change, you will have to coordinate roll-out

File: test/exoscale/vinyl/scan_test.clj [hunk 3]
```diff
@@ -29,14 +29,18 @@
   [x _]
   (inc x))
 
-;; For the sake of test run times we keep this small, bump at will
-;; in the REPL!
-(def small-test-data
-  (record-generator "small-bucket" 100))
+(defn inc-prefix
```
Comment (pyr) [inline test/exoscale/vinyl/scan_test.clj:32]: I am really happy to see the venerable `inc-prefix` go away from prod code

File: src/exoscale/vinyl/store.clj (hunk 2 above)
Comment (arnaudgeiser) [reply src/exoscale/vinyl/store.clj:508]: I don't think we'll have to coordinate any rollout.
What we need to do is just to bump the dependency on SOS and change `::store/marker` by `::store/continuation`.

https://github.com/exoscale/sos/pull/372/files#diff-69017ca6355bd8689b74b8391e43706cc7a7f113c5bafbbc9dae349ba1b80f81R225

**Review by pyr — APPROVED** [review]
Thanks @Fumesover, welcome to the Lisperati

**Pull request merged**